R/ChromBackendSpectra.R

```diff
@@ -0,0 +1,254 @@
+#' @importFrom methods callNextMethod
+setMethod("show", "ChromBackendSpectra", function(object) {
+    callNextMethod()
+    cat("\nThe Spectra object contains", length(object@spectra), "spectra\n")
+    if (object@inMemory) cat("\nPeaks data is cached in memory\n")
+})
+
+#' @rdname ChromBackendSpectra
+#' @note
+#' Does it needs to be a generic and implemented for ChromBackend ? It's very
+#' specific for this backend, also I would NOT allow replacement of it.
+#' @export
+chromSpectraIndex <- function(object) {
+    if (!is(object, "ChromBackendSpectra"))
+        stop("The object must be a 'ChromBackendSpectra' object.")
+    chromData(object, columns = "chromSpectraIndex", drop = TRUE)
+}
+
+#' @rdname hidden_aliases
+setMethod("factorize", "ChromBackendSpectra",
+          function(object, factorize.by = c("msLevel", "dataOrigin"),...) {
+              if (!all(factorize.by %in% chromVariables(object)))
+                  stop("All 'factorize.by' variables must be in chromData.")
+              if (!all(factorize.by %in% Spectra::spectraVariables(object@spectra)))
+                  stop("All 'factorize.by' variables must be in the ",
+                       "Spectra object.")
+              object@chromData$chromSpectraIndex <- factor(do.call(
+                  paste, c(object@chromData[, factorize.by], sep = "_")))
+              object@spectra$chromSpectraIndex <- factor(
+                  do.call(
+                      paste,
+                      c(as.list(Spectra::spectraData(object@spectra)[, factorize.by]),
+                        sep = "_")),
+                  levels = levels(chromSpectraIndex(object)))
+              object
+          })
+
+#' @rdname hidden_aliases
+#' @importMethodsFrom ProtGenerics backendParallelFactor
+setMethod("backendParallelFactor", "ChromBackendSpectra", function(object, ...) {
+    factor(chromSpectraIndex(object), levels = unique(chromSpectraIndex(object))) # now sure about that.
+})
+
+#' @rdname hidden_aliases
+#' @export
+setMethod("isReadOnly", "ChromBackendSpectra", function(object) TRUE)
+
+#' @rdname hidden_aliases
+setMethod("peaksData", "ChromBackendSpectra",
+          function(object, columns = peaksVariables(object),
+                   drop = FALSE, ...) {
+              if (object@inMemory || !length(object)) return(callNextMethod())
+              ## Ensure chromSpectraIndex only contains relevant levels needed
+              valid_levels <- chromSpectraIndex(object)
+              if (!all(levels(object@spectra$chromSpectraIndex) %in% valid_levels))
+                  object@spectra$chromSpectraIndex <- factor(object@spectra$chromSpectraIndex,
+                                                             levels = valid_levels)
+
+              ## Process peaks data
+              pd <- mapply(.process_peaks_data,
+                           cd = split(chromData(object), valid_levels),
+                           s = split(object@spectra, object@spectra$chromSpectraIndex),
```

Comment:
so, here use then the `fs` defined above for the split instead of `object@spectra$chromSpectraIndex`.